pkg/reconciler/apis/apibinding/apibinding_reconcile_permissionclaim.go

```diff
@@ -0,0 +1,305 @@
+			}
+		}
+		if !found {
+			claim, gvr, err := c.createPermissionClaimHelper(observedPC)
+			if err != nil {
+				return err
+			}
+			claims := removedPermissionClaims[gvr]
+			if claims == nil {
+				claims = []permissionClaim{}
+			}
+			claims = append(claims, claim)
+			removedPermissionClaims[gvr] = claims
+		}
+	}
+
+	klog.V(2).Infof("permission controller adding permission claims: %v, removing: %v and found invalid claims: %s", addedPermissionClaims, removedPermissionClaims, invalidClaims)
+
+	for gvr, claims := range addedPermissionClaims {
+		for _, claim := range claims {
+			// In the future, we will need to determine if the claim matches for any of the objects.
+			//TODO: use indexer instead of lister.
+			objs, err := claim.lister.Lister().List(labels.Everything())
+			if err != nil {
+				return err
+			}
+			for _, obj := range objs {
+				err = c.patchGenericObject(ctx, obj, claim, gvr, lc)
+				if err != nil {
+					return err
+				}
+			}
+			// If a here, lets add this to Observed Permission Claims
+			apiBinding.Status.ObservedAcceptedPermissionClaims = append(apiBinding.Status.ObservedAcceptedPermissionClaims, claim.claim)
+		}
+	}
+
+	for gvr, claims := range removedPermissionClaims {
+		for _, claim := range claims {
+			//TODO: use indexer instead of lister.
+			selector := labels.SelectorFromSet(labels.Set(map[string]string{claim.key: claim.label}))
+			objs, err := claim.lister.Lister().List(selector)
+			if err != nil {
+				return err
+			}
+			for _, obj := range objs {
+				err = c.patchGenericObject(ctx, obj, claim, gvr, lc)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// remove claims assume that all new ones were added.
+	apiBinding.Status.ObservedAcceptedPermissionClaims = removeClaims(apiBinding.Status.ObservedAcceptedPermissionClaims, removedPermissionClaims)
+
+	// Handle invalid claims
+	if len(invalidClaims) > 0 {
+		if identityMismatch {
+			conditions.MarkFalse(
+				apiBinding,
+				apisv1alpha1.PermissionClaimsAccepted,
+				apisv1alpha1.IdentityMismatchClaimInvalidReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Invalid AcceptedPermissionClaim. Please contact the APIExport owner to resolve or verify identity's bound.",
+			)
+		} else {
+			conditions.MarkFalse(
+				apiBinding,
+				apisv1alpha1.PermissionClaimsAccepted,
+				apisv1alpha1.UnknownPermissionClaimInvalidReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Claims have not been fully synced",
+			)
+		}
+		return fmt.Errorf("claims have not been synced")
+	}
+
+	conditions.MarkTrue(apiBinding, apisv1alpha1.PermissionClaimsAccepted)
+
+	return nil
+}
+
+func removeClaims(claims []apisv1alpha1.PermissionClaim, removedClaims mapGVRToClaim) []apisv1alpha1.PermissionClaim {
+
+	newClaims := []apisv1alpha1.PermissionClaim{}
+
+	for _, claim := range claims {
+		removed := false
+
+		for _, removedClaims := range removedClaims {
+			for _, c := range removedClaims {
+				if reflect.DeepEqual(c, claim) {
```

Comment:
`type PermissionClaim struct` has three fields which can be equated with a trivial:

```go
func (g GroupResource) Equal(other GroupResource) bool {
    return g.Group == other.Group && g.Resource == other.Resource
}

func (p PermissionClaim) Equal(other PermissionClaim) bool {
    return p.GroupResource == other.GroupResource && p.IdentityHash == other.IdentityHash
}
```

Can we please do this instead of using reflection, or semantic equality (which uses reflection)?

Reply:
yeah I had this, must have accidently dropped the commit sorry will fix